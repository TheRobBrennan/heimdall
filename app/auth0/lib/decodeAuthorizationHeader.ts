import { decodeAuth0JWT } from "./decodeAuth0JWT"

export const decodeAuthorizationHeader = async (
  authorizationHeaderValue: string
) => {
  try {
    // Use array destructuring to disregard the first item. We only care about the JWT.
    const [_, token] = authorizationHeaderValue.split("Bearer ")

    // DEBUG
    console.log(`
    [DEBUG] Auth0 decodeAuthorizationHeader received authorization header value:

    ${authorizationHeaderValue}
    `)

    // Decode and verify JWT using Auth0 keys
    const decodedAuth0JWT = await decodeAuth0JWT(token)
    console.log(
      `Decoded Auth0 JWT: ${JSON.stringify(decodedAuth0JWT, null, 2)}`
    )

    // Return the decoded JWT
    return decodedAuth0JWT
  } catch (e) {
    return console.error(
      `Unable to decode authorizationHeaderValue: ${authorizationHeaderValue}`
    )
  }
}
