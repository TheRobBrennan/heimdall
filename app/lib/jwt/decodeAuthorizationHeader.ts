import { decodeJWT } from "./decodeJWT"
import { decodeAuth0JWT } from "../../auth0/lib"

export const decodeAuthorizationHeader = async (authorizationHeader) => {
  try {
    // Use array destructuring to disregard the first item. We only care about the JWT.
    const [_, token] = authorizationHeader.split("Bearer ")

    // This decodes ANY well-formed JWT and does not vouch for validity or authenticity
    const decoded = decodeJWT(token)
    console.log(
      `Decoded generic JWT without token verification: ${JSON.stringify(
        decoded,
        null,
        2
      )}`
    )

    // Decode and verify JWT using Auth0 keys
    const decodedAuth0JWT = await decodeAuth0JWT(token)
    console.log(
      `Decoded Auth0 JWT: ${JSON.stringify(decodedAuth0JWT, null, 2)}`
    )
  } catch (e) {
    return console.error(
      `Unable to decode authorizationHeader: ${authorizationHeader}`
    )
  }
}
