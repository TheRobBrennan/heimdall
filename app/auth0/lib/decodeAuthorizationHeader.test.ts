import { cleanup } from "@testing-library/react"

import { decodeAuthorizationHeader as func } from "./decodeAuthorizationHeader"

// REVISIT: Update with a valid JWT token if you are running tests
export const VALID_JWT_TOKEN =
  "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6InNOTnN1RWZmU0lxeTRrcmtrQ2t4SSJ9.eyJodHRwczovL2hlaW1kYWxsLmlvL2p3dC9jbGFpbXMiOnsieC1oZWltZGFsbC1kZWZhdWx0LXJvbGUiOiJ1c2VyIiwieC1oZWltZGFsbC1hbGxvd2VkLXJvbGVzIjpbInVzZXIiXSwieC1oZWltZGFsbC11c2VyLWlkIjoiZ29vZ2xlLW9hdXRoMnwxMTYwNTg2NjgzMDIyOTA4NjE4MTAifSwiaXNzIjoiaHR0cHM6Ly9oZWltZGFsbC1kZW1vLnVzLmF1dGgwLmNvbS8iLCJzdWIiOiJnb29nbGUtb2F1dGgyfDExNjA1ODY2ODMwMjI5MDg2MTgxMCIsImF1ZCI6WyJodHRwczovL2hlaW1kYWxsLmlvL2RlbW8iLCJodHRwczovL2hlaW1kYWxsLWRlbW8udXMuYXV0aDAuY29tL3VzZXJpbmZvIl0sImlhdCI6MTYxMTQ2ODEwNCwiZXhwIjoxNjExNTU0NTA0LCJhenAiOiI2U25CUUw1WmQ3TjFvMHp5c3lmZnZhVUVGNXZoWUpFSCIsInNjb3BlIjoib3BlbmlkIHByb2ZpbGUifQ.oF9G3eB3lcScsneGpkG08hZDUP9WhLPhKnSsjpPKpo4PESRsRJxuyYGup9tSm1zi_nCJQVKqRf_DszYNHIIIfS2IfZhhxaHJhtzJOjiyY8ZS6Q2h-d_VyrXAuRahOwhJgBPwCk2UglGzhoHDgmqmNlZHT9bPTpJZMlu4XezDJG_PWDkZ3gurUPgeERS8gULI8YHFeKFZ10qkUbWqKP0QGLQIT66jamqJ-_AbUWPIXs8c0Y_l-7vrWKYQIZlI_HlismuZucTF6h-K2D9zpT_WoxeA5_xqY0riA1Dr1gbyifuUTn3jT3VxCEMSigA128FZavoXOOOhYvJzR7jeSsbqGQ"

describe("Our Auth0 decodeAuthorizationHeader lib", () => {
  afterEach(cleanup)
  describe('when supplied with a valid "Bearer <token>" string', () => {
    it("should receive a decoded JWT object", async () => {
      const mockHttpAuthorizationHeaderValue = `Bearer ${VALID_JWT_TOKEN}`
      const result = await func(mockHttpAuthorizationHeaderValue)
      expect(result).toMatchSnapshot()
    })
  })
  describe('when supplied with a string that does not contain a valid "Bearer <token>"', () => {
    it("should return undefined", async () => {
      const mockHttpAuthorizationHeaderValue = `Bearer fakeToken`
      const result = await func(mockHttpAuthorizationHeaderValue)
      expect(result).toEqual(undefined)
    })
  })
  describe("when supplied with an unexpected string", () => {
    it("should return undefined", async () => {
      const mockHttpAuthorizationHeaderValue = `Here is an example string that is neither a JWT token nor a valid string in the HTTP Authorization header`
      const result = await func(mockHttpAuthorizationHeaderValue)
      expect(result).toEqual(undefined)
    })
  })
})
