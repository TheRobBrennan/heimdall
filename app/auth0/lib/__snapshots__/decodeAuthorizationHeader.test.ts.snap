// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Our Auth0 decodeAuthorizationHeader lib when supplied with a valid "Bearer <token>" string should receive a decoded JWT object 1`] = `
Object {
  "aud": Array [
    "https://heimdall.io/demo",
    "https://heimdall-demo.us.auth0.com/userinfo",
  ],
  "azp": "6SnBQL5Zd7N1o0zysyffvaUEF5vhYJEH",
  "exp": 1611951927,
  "https://heimdall.io/jwt/claims": Object {
    "x-heimdall-allowed-roles": Array [
      "user",
    ],
    "x-heimdall-default-role": "user",
    "x-heimdall-user-id": "google-oauth2|116058668302290861810",
  },
  "iat": 1611865527,
  "iss": "https://heimdall-demo.us.auth0.com/",
  "scope": "openid profile",
  "sub": "google-oauth2|116058668302290861810",
}
`;
